File: symfony/src/Entity/Content/Article.php
<?php

namespace App\Entity\Content;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Annotation\Groups;
use App\Repository\Content\ArticleRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use App\Entity\Players\Player;
use App\Entity\Cards\Deck;

#[ORM\Entity(repositoryClass: ArticleRepository::class)]
#[ORM\Table(name: 'article')]
class Article
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    #[Groups(['article:read'])]
    private ?int $id = null;

    #[ORM\Column(type: 'string', length: 300)]
    #[Groups(['article:read', 'article:write'])]
    private string $title = '';

    #[ORM\Column(type: 'string', length: 300)]
    #[Groups(['article:read', 'article:write'])]
    private string $slug = '';

    #[ORM\Column(type: 'text')]
    #[Groups(['article:read', 'article:write'])]
    private string $body = '';

    #[ORM\Column(type: 'text', nullable: true)]
    #[Groups(['article:read', 'article:write'])]
    private ?string $excerpt = null;

    #[ORM\Column(type: 'string', length: 200, nullable: true)]
    #[Groups(['article:read', 'article:write'])]
    private ?string $coverImageUrl = null;

    #[ORM\Column(type: 'string', length: 20)]
    #[Groups(['article:read', 'article:write'])]
    private string $status = 'Draft';

    #[ORM\Column(type: 'string', length: 20)]
    #[Groups(['article:read', 'article:write'])]
    private string $articleType = 'Guide';

    #[ORM\Column(type: 'integer')]
    #[Groups(['article:read', 'article:write'])]
    private int $viewCount = 0;

    #[ORM\Column(type: 'datetime', nullable: true)]
    #[Groups(['article:read', 'article:write'])]
    private ?\DateTimeInterface $publishedAt = null;

    #[ORM\Column(type: 'datetime', nullable: true)]
    #[Groups(['article:read', 'article:write'])]
    private ?\DateTimeInterface $createdAt = null;

    #[ORM\Column(type: 'datetime', nullable: true)]
    #[Groups(['article:read', 'article:write'])]
    private ?\DateTimeInterface $updatedAt = null;

    #[ORM\ManyToOne(targetEntity: Player::class, inversedBy: 'articles')]
    #[ORM\JoinColumn(nullable: false)]
    private ?Player $author = null;

    #[ORM\ManyToOne(targetEntity: Deck::class, inversedBy: 'articles')]
    #[ORM\JoinColumn(nullable: true)]
    private ?Deck $featuredDeck = null;

    #[ORM\ManyToOne(targetEntity: ArticleComment::class, inversedBy: 'article')]
    #[ORM\JoinColumn(nullable: false)]
    private ?ArticleComment $comments = null;

    #[ORM\ManyToMany(targetEntity: ArticleTag::class)]
    #[ORM\JoinTable(name: 'article_tags_m2m')]
    private Collection $tags;

    public function __construct()
    {
        $this->tags = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getTitle(): string
    {
        return $this->title;
    }

    public function setTitle(string $title): static
    {
        $this->title = $title;
        return $this;
    }

    public function getSlug(): string
    {
        return $this->slug;
    }

    public function setSlug(string $slug): static
    {
        $this->slug = $slug;
        return $this;
    }

    public function getBody(): string
    {
        return $this->body;
    }

    public function setBody(string $body): static
    {
        $this->body = $body;
        return $this;
    }

    public function getExcerpt(): ?string
    {
        return $this->excerpt;
    }

    public function setExcerpt(?string $excerpt): static
    {
        $this->excerpt = $excerpt;
        return $this;
    }

    public function getCoverImageUrl(): ?string
    {
        return $this->coverImageUrl;
    }

    public function setCoverImageUrl(?string $coverImageUrl): static
    {
        $this->coverImageUrl = $coverImageUrl;
        return $this;
    }

    public function getStatus(): string
    {
        return $this->status;
    }

    public function setStatus(string $status): static
    {
        $this->status = $status;
        return $this;
    }

    public function getArticleType(): string
    {
        return $this->articleType;
    }

    public function setArticleType(string $articleType): static
    {
        $this->articleType = $articleType;
        return $this;
    }

    public function getViewCount(): int
    {
        return $this->viewCount;
    }

    public function setViewCount(int $viewCount): static
    {
        $this->viewCount = $viewCount;
        return $this;
    }

    public function getPublishedAt(): ?\DateTimeInterface
    {
        return $this->publishedAt;
    }

    public function setPublishedAt(?\DateTimeInterface $publishedAt): static
    {
        $this->publishedAt = $publishedAt;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    public function setCreatedAt(?\DateTimeInterface $createdAt): static
    {
        $this->createdAt = $createdAt;
        return $this;
    }

    public function getUpdatedAt(): ?\DateTimeInterface
    {
        return $this->updatedAt;
    }

    public function setUpdatedAt(?\DateTimeInterface $updatedAt): static
    {
        $this->updatedAt = $updatedAt;
        return $this;
    }

    #[Groups(['article:read'])]
    public function getAuthorId(): ?int
    {
        return $this->author?->getId();
    }

    public function getAuthor(): ?Player
    {
        return $this->author;
    }

    public function setAuthor(?Player $author): static
    {
        $this->author = $author;
        return $this;
    }

    #[Groups(['article:read'])]
    public function getFeaturedDeckId(): ?int
    {
        return $this->featuredDeck?->getId();
    }

    public function getFeaturedDeck(): ?Deck
    {
        return $this->featuredDeck;
    }

    public function setFeaturedDeck(?Deck $featuredDeck): static
    {
        $this->featuredDeck = $featuredDeck;
        return $this;
    }

    #[Groups(['article:read'])]
    public function getCommentsId(): ?int
    {
        return $this->comments?->getId();
    }

    public function getComments(): ?ArticleComment
    {
        return $this->comments;
    }

    public function setComments(?ArticleComment $comments): static
    {
        $this->comments = $comments;
        return $this;
    }

    public function getTags(): Collection
    {
        return $this->tags;
    }

    public function addTags(ArticleTag $item): static
    {
        if (!$this->tags->contains($item)) {
            $this->tags->add($item);
        }
        return $this;
    }

    public function removeTags(ArticleTag $item): static
    {
        $this->tags->removeElement($item);
        return $this;
    }

}

File: spring/src/main/java/cardsproject/domain/content/Article.java
package cardsproject.domain.content;

import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "articles")
public class Article {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String title = "";
    private String slug = "";
    private String body = "";
    private String excerpt;
    private String coverImageUrl;
    @Enumerated(EnumType.STRING)
    private ArticleStatusType status;
    @Enumerated(EnumType.STRING)
    private ArticleArticleTypeType articleType;
    @Enumerated(EnumType.STRING)
    private ArticleLanguageType language;
    private Integer viewCount = 0;
    private Integer likesCount = 0;
    private Boolean isFeatured = false;
    private LocalDateTime publishedAt;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @Column(name = "author_id")
    private Long authorId;
    @Column(name = "featured_deck_id")
    private Long featuredDeckId;

    // M2M: tags managed via join table

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    public String getSlug() { return slug; }
    public void setSlug(String slug) { this.slug = slug; }
    public String getBody() { return body; }
    public void setBody(String body) { this.body = body; }
    public String getExcerpt() { return excerpt; }
    public void setExcerpt(String excerpt) { this.excerpt = excerpt; }
    public String getCoverImageUrl() { return coverImageUrl; }
    public void setCoverImageUrl(String coverImageUrl) { this.coverImageUrl = coverImageUrl; }
    public ArticleStatusType getStatus() { return status; }
    public void setStatus(ArticleStatusType status) { this.status = status; }
    public ArticleArticleTypeType getArticleType() { return articleType; }
    public void setArticleType(ArticleArticleTypeType articleType) { this.articleType = articleType; }
    public ArticleLanguageType getLanguage() { return language; }
    public void setLanguage(ArticleLanguageType language) { this.language = language; }
    public Integer getViewCount() { return viewCount; }
    public void setViewCount(Integer viewCount) { this.viewCount = viewCount; }
    public Integer getLikesCount() { return likesCount; }
    public void setLikesCount(Integer likesCount) { this.likesCount = likesCount; }
    public Boolean getIsFeatured() { return isFeatured; }
    public void setIsFeatured(Boolean isFeatured) { this.isFeatured = isFeatured; }
    public LocalDateTime getPublishedAt() { return publishedAt; }
    public void setPublishedAt(LocalDateTime publishedAt) { this.publishedAt = publishedAt; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    public Long getAuthorId() { return authorId; }
    public void setAuthorId(Long authorId) { this.authorId = authorId; }
    public Long getFeaturedDeckId() { return featuredDeckId; }
    public void setFeaturedDeckId(Long featuredDeckId) { this.featuredDeckId = featuredDeckId; }

    // ── Business operations ──────────────────────────────────────────
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void publish() {
        // TODO: implement publish
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void archive() {
        // TODO: implement archive
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void incrementView() {
        // TODO: implement incrementView
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void like() {
        // TODO: implement like
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void unlike() {
        // TODO: implement unlike
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public Integer readingTimeMinutes() {
        // TODO: implement readingTimeMinutes
        return null;
    }

    // ── Validation rules ─────────────────────────────────────────────
    @jakarta.validation.constraints.AssertTrue(message = "Article view count must not be negative")
    @com.fasterxml.jackson.annotation.JsonIgnore
    public boolean isViewCountNotNegativeValid() {
        return (getViewCount() == null || getViewCount() >= 0);
    }
    @jakarta.validation.constraints.AssertTrue(message = "Article likes count must not be negative")
    @com.fasterxml.jackson.annotation.JsonIgnore
    public boolean isLikesCountNotNegativeValid() {
        return (getLikesCount() == null || getLikesCount() >= 0);
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static final java.util.Map<ArticleStatusType, java.util.List<ArticleStatusType>> ALLOWED_TRANSITIONS =
        java.util.Map.ofEntries(
        java.util.Map.entry(ArticleStatusType.DRAFT, java.util.List.of(ArticleStatusType.PUBLISHED)),
        java.util.Map.entry(ArticleStatusType.PUBLISHED, java.util.List.of(ArticleStatusType.ARCHIVED)),
        java.util.Map.entry(ArticleStatusType.ARCHIVED, java.util.List.of(ArticleStatusType.DRAFT))
        );

    public void assertTransition(ArticleStatusType to) {
        java.util.List<ArticleStatusType> allowed = ALLOWED_TRANSITIONS.getOrDefault(this.getStatus(), java.util.List.of());
        if (!allowed.contains(to)) {
            throw new IllegalStateException("Transition " + this.getStatus() + " -> " + to + " not allowed");
        }
    }
}
